### ISSUE TITLE
MDBX Freelist Issues

### ISSUE THREAD

**Issue by shekhirin** [paradigmxyz/reth#5228]
## Problem

MDBX stores data in pages. Roughly speaking, there's a list of active pages currently used by a B-Tree, and a list of free pages that can be used without an allocation of new pages.

When a new key-value record is written, the value can be larger than a page size, thus requiring overflow pages that will be placed so that there would be a consecutive read from the disk instead of random one. As said before, it can be beneficial to check the list of free pages first and try to find a sequence of pages needed for such record there, instead of allocating new pages. If there are not enough consecutive pages of required length in the freelist, the database allocates new pages.

There are multiple sources of freelist growth:
- Deleting or updating data
   - When a record is deleted, the pages that are no longer in use are marked as freed and "placed" into a freelist for further reclamation by new records.
   - The same happens when the record is updated and its size is decreased – no longer needed overflow pages are marked as freed.
   - Frequent scenario for Reth when running as a pruned/full node. That's why we have deletion limits for the pruner: https://github.com/paradigmxyz/reth/pull/4880.
- Reading data in one transaction, and modifying it in the other
   - Due to MVCC, read transactions point at the snapshot of data at the start of transaction. So when other write transactions modify the same data, new "version" of this data is created with an updated value. But when the read transaction is closed and the snapshot of outdated data is no longer being read by any transaction, the pages are marked as freed.
   - So, if you have long read transactions that access some frequently modified data, they will accumulate a lot of pages of old data that will increase the size of freelist.
   - Also, read transactions block the recycling of freelist, so its size will not decrease until read transactions are closed.
   - Might be a frequent scenario for Reth if we have dangling/stuck read transactions (metrics should help to identify them better https://github.com/paradigmxyz/reth/pull/5149)
- Transaction rollback
   - When the data is modified inside a write transaction, but then it gets aborted/reverted, the pages of modified data are marked as freed.
   - Not really a frequent scenario for Reth, only on reorgs.
- Random writes
   - If one transaction does 10 small appends to the same table, all of them will go to the same B-Tree page. It will be just 1 CoW (copy-on-write of database page). No need to make 10 CoWs – readers will not see uncommitted updates.
   - When insertions are random, more B-Tree pages are modified, hence more CoWs are done.
   - 1 CoW is a CoW of the entire b+tree branch (from the leaf being updated to the root). Which means updating 2 co-located pages updates fewer pages than updating 2 random pages. Fewer updated pages means less pages going into the freelist.
   - Frequent scenario for Reth when hash insertions are made, because they can't be ordered in most of the cases (e.g. Merkle sync stage). Same for Execution stage, because storage slots and accounts are scattered all across the database.
- Some other more rare scenarios that I believe not so relevant for Reth.

The problems arise when we want to insert a huge value (say, 10s of kilobytes of data that need 10-20 overflow pages given a page size of 4KB). MDBX searches through all of the freelist, trying to find a long enough sequence of free pages, and if it doesn't succeed, allocates new. This search can take up to 5-15 seconds with the database of our size.

## Solution

There's no easy fix to this problem, as it's inherent to MDBX. Erigon also struggles with this issue (https://github.com/ledgerwatch/erigon/pull/7798, https://github.com/ledgerwatch/erigon/issues/8224). But some mitigations can be implemented.

### Page size increase

Currently, we use 4KB as a minimum page size and try to match it with the page size of OS (which is 4KB in most of the cases). If we increase the page size 2x to 8KB (https://github.com/paradigmxyz/reth/pull/5163), it will mainly:
- Allocate less overflow pages because large values will more likely fit into one page
- Reduce the size of freelist, thus improving the search on inserts of large values

But it comes at a cost: not matching the OS page size means more I/O (Erigon estimates it at 10% more).

### Prevent overflow pages by splitting long records

If we insert a value larger than the page size, MDBX will use an overflow page, and maybe search through a freelist. But if we split the records with long values on our own (say, turning Bytecodes/Receipts/Transactions into DupSort tables), we will always use just one page per record.

But it comes at a cost: we need to do more maintenance to support records that are split into multiple rows, and re-do how we read these values. The good thing is that we have good enough DB abstractions to experiment with that.

### Prevent overflow pages by writing large records into a separate storage

Basically, it's snapshots. The large records doesn't even enter MDBX, and are placed directly into static files, thus making the insertion and deletion (say, on reorgs) fast.

But it comes at a cost: we need to do ETL, and write data from the initial sync directly into static files, avoiding the MDBX. Also, we need to think about atomicity and guarantees.

### Change database

Do other databases have this issue? Is it inherent to a use of B-Tree, or maybe related to mmap in any way? We need to investigate it better.

**Comment by AskAlexSharov** [comment]
Increasing pageSize solving this issue from multiple directions:  
- less overflow pages
- less metadata (freelist) to manage
- mdbx also trying to avoid overflow pages in freelist - and splits it to many 1-page-records (each page storing list of free pageID’s) reducing probability of finding n-page-sequences in freelist. But if pageSize is bigger - it can store more pageID’s on 1-page-record - improving probabilities and reducing fragmentation.
- reduce probability of page-split/page-merge 
- b+tree depth getting smaller - less branch pages to  update

- overall less db fragmentation (b+tree gravitating towards random pages distribution) and more compressed-fs friendly. 
- some systems (like modern macos) using 16kb pagesize by default.
- yes, 10% write io increase on chain tip, but this 10% is co-located with existing writes (not additional random pages), seems it didn’t impact commit speed. 
- reads - b+tree depth getting smaller, but b+tree leafs getting bigger (binary search is there). Seems no big difference.

**Comment by AskAlexSharov** [comment]
Etcd - RAFT replicated db - using “boltdb” under the hood. And boltdb is the mdbx re-written in golang - with some features loss - and even more simplified freelist - it does copy whole FreeList on each commit: https://hn.svelte.dev/item/30013919

**Comment by AskAlexSharov** [comment]
Oh, and 4kb pagesize - has 8tb db limit. Erigon’s version of archive Polygon is 12Tb now.

**Comment by sandakersmann** [comment]
Introducing Firewood: A Next-Generation Database Built for High-Throughput Blockchains

https://www.avax.network/blog/introducing-firewood-a-next-generation-database-built-for-high-throughput-blockchains

**Comment by yilongli** [comment]
> ### Change database
> Do other databases have this issue? Is it inherent to a use of B-Tree, or maybe related to mmap in any way? We need to investigate it better.

IIUC, you are saying that the freelist is literally implemented as an on-disk linked list of free page IDs? If so, the root cause is the inefficient data structure of freelist; I don't see how the problem is inherent to B-tree or mmap-based buffer pool.

**Comment by yilongli** [comment]
The problem of freelist appears to be very similar to splitting and coalescing memory chunks in malloc. I am sure there are much better data structures out there.

However, I am still not sure that (ab)using MDBX large values in Reth is a good idea in the first place. What are the typical access patterns of these large values? For read-only records, since they don't contribute to the growth of freelist, just leave them be. For records that can be updated, I am worried that using large values may aggravate the problem of write amplification in vanilla B-tree. Do you know if mdbx will write multiple pages when only a small portion of a large value is changed? If so, I would say it's probably better to break them up into smaller values.

**Comment by sevazhidkov** [comment]
> MDBX searches through all of the freelist, trying to find a long enough sequence of free pages, and if it doesn't succeed, allocates new.
> 

I would like to elaborate on why MDBX spends a significant amount of time searching for a free page sequence. There are separate concepts of a `freelist` (referred to as GC in the MDBX codebase) and a `reclaimed-list` (known as `re-list` in the codebase). 

When pages become available (for the reasons described in the original issue), they are added to the `freelist`. However, when determining where to place a new value, MDBX actually looks in the `reclaimed-list`. Only if it doesn’t find a place there, it starts looking in the `freelist`.

So, when searching for a large sequence of pages that's not immediately found in the `reclaimed-list`, MDBX sequentially goes through the `freelist`, extracting freed pages and adding them to the `reclaimed-list`. After each `freelist` entry is processed and added to the `reclaimed-list`, it searches the `reclaimed-list` for a sequence again. Essentially, it incrementally builds a larger `reclaimed-list` with each iteration, in hopes of eventually finding the desired sequence within it.

```python
# This is pseudocode interpretation of page_alloc_slowpath, leaving out some of the details

page_sequence = find_consecutive_page_sequence(reclaimed_list, n_pages)

if not page_sequence:
    # Depending on FIFO or LIFO setting, this iteration either goes
    for item in freelist[last_processed_freelist_item:]:
        # Mark pages from this GC entry as available for usage
        reclaimed_list.extend(item.pages)
        # In the next run, we shouldn't go over this freelist element again
        last_processed_freelist_item += 1
        # Can we find a sequence now?
        page_sequence = find_consecutive_page_sequence(reclaimed_list, n_pages)
        if page_sequence:
            # allocate at this page sequence
            break
        if len(reclaimed_list) > rp_augment_limit:
            # allocate at the end of the map
            break

# * reclaimed_list and last_processed_freelist_item are dropped 
# at the end of transaction
# * freelist is saved in the database indefinitely
```

If MDBX has scanned all the `freelist` entries and still hasn't found the required sequence, or if it has reached the `rp_augment_limit` (which restricts the size of the reclaimed-list), it then allocates a sequence from the end of the map.

My theory, supported by tracing of [this repo](https://github.com/shekhirin/mdbx-freelist), is that the huge `mdbx_put` latencies are pretty much executing thousands of iterations in this loop.

Fundamentally, this algorithm has a few issues related to the usage in reth:

1. During every iteration of the `freelist` loop, the `reclaimed-list` search is repeated. Also, on each iteration, time is spent appending page numbers to the `reclaimed-list` and maintaining it as a sorted array.
2. The `freelist` is processed in sequential order. This approach does not attempt to find consecutive pages in the `freelist` to add to the reclaimed list sooner than their sequential order. Instead, it merely adds the oldest or newest pages from the `freelist`, hoping they will eventually form a consecutive sequence.
3. Fortunately, if someone is making multiple puts in the same transaction, MDBX won't go over the `freelist` twice. It would preserve the collected `reclaimed-list`. However, the `reclaimed-list` is stored in the transaction object, not in the database file. So, when new transaction is created, `reclaimed-list` is wiped out and has to be built from `freelist` again.

So what can be done, while keeping all data in MDBX [1] and not splitting the values at the application level?

**Slab allocation (ballast)**

We can pre-allocate large slab objects and then delete them individually before writing large transaction blobs. The size of the pre-allocated slab object must be greater than or equal to the size of the object being written, measured in pages. This approach would guarantee a series of consecutive pages in the database to write a value.

However, deleting a large object puts free pages to the `freelist`, not the `reclaimed-list` (as any other deletions). How do we get the freed up sequence to `reclaimed-list` fast, so it can be used by a new put? MDBX needs to start the `freelist` iteration from the most recent values in the freelist (LIFO). This is a MDBX setting: **[`MDBX_LIFORECLAIM`](https://erthink.github.io/libmdbx/group__c__opening.html#gga9138119a904355d245777c4119534061a37585486d4e99bcb599cc15b408137b0)**. If that’s enabled, the recent values will quickly be placed in the `reclaimed`-list and will form a needed consecutive sequence.

We don’t have to always use LIFO. It can be enabled just for the transactions with big blobs, when we need to catch the freed up slab pages, and keep it disabled for all other scenarios. 

**Tuning `rp_augment_limit` parameter**

`rp_augment_limit` sets a limit on the number of pages in the `reclaimed-list`. If the list reaches this limit, MDBX will not attempt to add new elements by iterating over the `freelist`. Instead, MDBX will append new values to the end of the map, either by creating new pages or expanding the database file. Reducing this parameter can significantly improve performance because iterating over the `freelist`, which is time-consuming for large page writes, is avoided.

If we determine that transaction data blobs are unlikely to find a place on the freelist before overflowing the `reclaimed`-list (this can be verified on a node by checking the page numbers of newly allocated transactions), we can short-circuit this process to avoid unnecessary work. By setting a small limit on the `reclaimed-list` using `rp_augment_limit`, we ensure that we don't spend too much time looking at the `freelist` for the large writes. As a result, all large transaction blobs will be placed at the end of the current map.

This approach has a drawback: it will always allocate new pages for large transaction blobs. However, if the alternative is to simply store the blobs on disk, this approach would essentially do the same.

**Proposal**

I would suggest the following plan, assuming we want to keep it simple with just mdbx, not filesystem:

- Implement a check to determine if a block contains a transaction that's too large to place immediately and test it for correctness.
- Identify the page numbers where these large transactions end up.
- If these are coming from the end of the database (which seems likely), apply `rp_augment_limit` to these blocks. This way, the most likely scenario would only require a config change and some testing.
- If these are coming from the middle of the database, implement slab allocation.

—

[1] Putting this on disk instead is probably fine. In this message, I'm just assuming reth is going with MDBX all-in for simplicity.

**Comment by 0xprames** [comment]
fwiw erigon also seems to have touched up `rp_augment_limit` [here](https://github.com/ledgerwatch/erigon/pull/1465/files) as a response to: https://github.com/erthink/libmdbx/issues/158

**Comment by 0xprames** [comment]
i would like to understand why #5163 is a blocker to resolve this, is the canonical commit time increase "too much"

and what does moving to another db look like in terms of the requirements for the db?

**Comment by 0xprames** [comment]
>  I don't see how the problem is inherent to B-tree or mmap-based buffer pool.

it doesn't seem like its inherent to either of these things, but rather how mdbx leverages the freelist. i think the b-tree/mmap was mentioned as an investigative question in the original issue

**Comment by yilongli** [comment]
> > I don't see how the problem is inherent to B-tree or mmap-based buffer pool.
> 
> it doesn't seem like its inherent to either of these things, but rather how mdbx leverages the freelist. i think the b-tree/mmap was mentioned as an investigative question in the original issue

Yep the inefficient implementation of freelist is not an inherent problem to the use of B-tree as an index structure. However, one could argue that LSM doesn't suffer from this problem because it requires a periodic background compaction process to work in the first place.

**Comment by yilongli** [comment]
@sevazhidkov Thanks for the detailed analysis! I didn't realize that freelist is implemented in such an inefficient way. Basically, the `reclaimed-list` is an ephemeral data structure that is only used in a transaction (rebuilt and discarded in every txn).

However, I am not yet convinced what you suggested would solve the problem. Note that the current search process, despite very inefficient, does allow you to coalesce adjacent memory chunks. This is critical to avoid external fragmentation. Trying to short-circuit this process using heuristics might result in low space utilization or even unbounded space growth in the presence of pathological workloads.

More specifically, how do you prevent a large slab from being splitted up by single-page requests? And what if the distribution of the size of large values is dispersed? e.g., how do you avoid a 10-page slab being used by a 2-page request? This seems difficult, if not impossible, to get right if we want to keep the `freelist` as it is. Heuristics might get you right most of time, but this is not enough for a long-running instance like Reth.

**Comment by sevazhidkov** [comment]
> This is critical to avoid external fragmentation. Trying to short-circuit this process using heuristics might result in low space utilization or even unbounded space growth in the presence of pathological workloads.

I agree that this process should be avoided generally. I second your questions about thinking through data access patterns and figuring out the exact right storage. I’ve heard that, at some point in the future, transactions would be added to append-only snapshots instead of mdbx (@shekhirin is that correct?). 

However, as an immediate solution: 

* The process should only be short-circuited on the problematic workflow: inserting transaction data blobs. My assumption is that there are no many similar writes in reth. All of the other writes, including blocks with smaller transactions will still work on recycling used pages. 
* My theory is that MDBX might be hitting the `reclaimed-list` limit most of the time, when experiencing high latencies. If that’s the case, the solution would not be worse that the  status quo fragmentation-wise and much better performance-wise. 
* We don’t have to radically reduce this parameter to zero. 10x reduction would still save a lot of recycling iterations, but improve 

>More specifically, how do you prevent a large slab from being splitted up by single-page requests?

We can delete it right before writing a large transaction data blob, in the same transaction, as an immediately preceding action. 

>And what if the distribution of the size of large values is dispersed?

The slab allocation normally deals with this by allocating objects of different sizes in buckets, and freeing up the smallest object that’s larger the the object being written. It's true that we'll have to come up with some arbitrary bucket parameters, but if we're thinking about transaction data, this is very analyzeable.

**Comment by shekhirin** [comment]
> Heuristics might get you right most of time, but this is not enough for a long-running instance like Reth.

good thing about Reth write workload is that it's very predictable and happens only once every 12s (on Mainnet), so IMO we can afford such optimizations tailored for a use case like this

> I’ve heard that, at some point in the future, transactions would be added to append-only snapshots instead of mdbx

right, and the plan is to not insert into the database any data that'll be retired later, see https://github.com/paradigmxyz/reth/issues/4513

**Comment by yorickdowne** [comment]
A DB that was written specifically to store a Ethereum state is Paprika. https://github.com/NethermindEth/Paprika

Design document is linked from its readme

**Comment by yilongli** [comment]
> * The process should only be short-circuited on the problematic workflow: inserting transaction data blobs. My assumption is that there are no many similar writes in reth. All of the other writes, including blocks with smaller transactions will still work on recycling used pages.

If the problematic writes can be isolated cleanly, that certainly simplifies the problem a lot.

> * My theory is that MDBX might be hitting the `reclaimed-list` limit most of the time, when experiencing high latencies. If that’s the case, the solution would not be worse that the  status quo fragmentation-wise and much better performance-wise.

But we don't know if reducing this threshold will cause other requests that would've succeeded to fail. Thus, I don't think this heuristic is guaranteed to be strictly better than the current situation.

> The slab allocation normally deals with this by allocating objects of different sizes in buckets, and freeing up the smallest object that’s larger the the object being written. It's true that we'll have to come up with some arbitrary bucket parameters, but if we're thinking about transaction data, this is very analyzeable.

Exactly, so a more robust solution is to use multiple freelists, one for each slab size. If the workloads are simple/predictable enough, we may not need to implement coalescing of adjacent chunks, so freelists can remain as unordered linked lists.

**Comment by 0xprames** [comment]
> A DB that was written specifically to store Ethereum state and storage is Paprika. https://github.com/NethermindEth/Paprika

Design document is linked from its readme

awesome share @yorickdowne

**Comment by yorickdowne** [comment]
It’s an interesting gambit by the Nethermind team @0xprames . They spend so much time trying to press RocksDB into service - like the mdbx discussions here - they figured they may as well write a storage engine optimized specifically for Ethereum.

This could end up with them spending much more time on storage maintenance; or less / the same but with an optimized backend now.

Worth a close look I figure.

**Comment by michaelsproul** [comment]
Have you considered Postgres? It's been on my TODO list to try as a backend for Lighthouse or the slasher, as it seems every major company & many database academics recommend it. E.g. in the Nethermind Paprika docs they link to a [paper by the CMU database group](https://db.cs.cmu.edu/mmap-cidr2022/) which advises _against_ `mmap` and in favour of old-fashioned DBMS. I think in Postgres you could [`VACUUM`](https://www.postgresql.org/docs/current/sql-vacuum.html) between blocks to reclaim space.

**Comment by yorickdowne** [comment]
Interesting idea. zkSync Era btw does Postgres plus RocksDB, both.

**Comment by 0xMelkor** [comment]
Hi guys, I'm experimenting on rocksdb. The storage crate is flexible enough to introduce it.
I'm using [rust-rocksdb](https://github.com/rust-rocksdb/rust-rocksdb) and the work proceeds well.

Feel free to share your thoughts if needed, I'm always open for collaboration.
Cheers

**Comment by 0xAlcibiades** [comment]
I'd posit here to look at the journey ceph had, where they ended up using rocksdb for journaling and a raw block device with bluestore for object storage, implementing a lightweight filesystem in bluestore. They are now looking at stuff like spdk and dpdk, io_uring to bring the IO up into userland. The intermediary step for them in the filestore object storage daemon was to indeed use rocksdb as a keyvalue store inside of an existing filesystem such as xfs/btrfs. There are also some rocksdb tunings which may be applied.

Fast keyvalue stores are hard, there are some other options out there to consider like rondb.
